### PR TITLE
[Intel GPU] Add XPU memory-related APIs

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129919
* #130923

# Motivation
According to https://github.com/pytorch/pytorch/issues/116322, we will help unify the device allocator. So we introduce a simple xpu device allocator only with the key functionality first. And expect to add some memory statistics-related functionality after the unification. 
But now, some memory statistic-related APIs listed in https://github.com/pytorch/pytorch/issues/127929 are requested. We need more time to unify the device allocator. In order to facilitate the user experience, we expect to support these memory statistic-related APIs before the unification.

# Additional Context
Fixes: #127929

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

File: c10/xpu/XPUCachingAllocator.cpp
Comment (dvrogozh) [inline c10/xpu/XPUCachingAllocator.cpp:41]:
```suggestion
      "Negative tracked stat in XPU allocator (likely logic error).");
```

**Comment by dvrogozh** [comment]
Add to PR description to automatically close the issue on merge:
```
Fixes: #127929
```

File: c10/xpu/XPUCachingAllocator.cpp
Comment (guangyey) [reply c10/xpu/XPUCachingAllocator.cpp:41]: updated. Thanks.

**Review by dvrogozh — APPROVED** [review]
looks good

File: c10/xpu/XPUCachingAllocator.cpp
Comment (EikanWang) [inline c10/xpu/XPUCachingAllocator.cpp:1]: Most of the changes are the same as the CUDA Caching Allocator. Should we abstract these utility functions and reuse these functions for both CUDA and XPU?

File: c10/xpu/XPUCachingAllocator.h
Comment (EikanWang) [inline c10/xpu/XPUCachingAllocator.h:35]: ditto

File: c10/xpu/XPUCachingAllocator.cpp
Comment (guangyey) [reply c10/xpu/XPUCachingAllocator.cpp:1]: I will unify the device allocator for `cuda` and `xpu` in a separate PR.

**Comment by dvrogozh** [comment]
What's the ETA to get this merged? These APIs block running some HF tests which might help to uncover other important issues.

**Review by abhilash1910 — APPROVED** [review]
LGTM! This was indeed needed. Thanks @guangyey

**Comment by dvrogozh** [comment]
@guangyey : can you, please, update status of this PR? When it will be merged? If not, what's pending? As of now it's not clear from the comments above. You plan some refactoring done after this PR merged, but it seems no actions taken to actually merge it. Maintainers are not pinged for review.

**Comment by guangyey** [comment]
Hi @dvrogozh , this PR is ready to review. I think Alban will review it when it is in his queue of regular PRs.

**Comment by dvrogozh** [comment]
@albanD : can you, please, help review this PR? that's long standing PR with key uAPIs heavily used across ecosystem projects.

**Comment by EikanWang** [comment]
@guangyey , could you help rebase the code and pass the CI?

**Review by gujinghui — APPROVED** [review]

**Comment by dvrogozh** [comment]
@albanD : can this PR, please, be reviewed? It would be nice to have it in upcoming pytorch 2.5 release since it blocks running some workloads and tests with huggingface.

**Comment by albanD** [comment]
This is pretty hard to review since I am not familiar with this code. I guess this is copy pasted from the cuda side?
What is the plan to align these APIs with other devices?

**Comment by guangyey** [comment]
> This is pretty hard to review since I am not familiar with this code. I guess this is copy pasted from the cuda side?

Yes, this code logic is the same as cuda.

**Comment by guangyey** [comment]
> What is the plan to align these python APIs with other devices?

In long term, we intend to give a generic python memory-related API design [[RFC] A device-agnostic Python device memory related API design for stream-based accelerators](https://github.com/pytorch/pytorch/issues/134978) on top of https://github.com/pytorch/pytorch/issues/116322.
On summary is 
- abstract C++ device caching allocator
- unify the C++ API via the abstract, like `GetAllocator(device_type)->emptyCache()`
- unify the python frontend API via C++ API, like `torch.empty_cache`.

**Review by EikanWang — APPROVED** [review]

**Comment by EikanWang** [comment]
> This is pretty hard to review since I am not familiar with this code. I guess this is copy pasted from the cuda side? What is the plan to align these python APIs with other devices?

Hi @albanD , As @guangyey mentioned, this logic and implementation is similar with CUDA. And this PR intends to serve HuggingFace Intel GPU support because some HuggingFace models leverage these APIs to get device memory information. 

Regarding python APIs alignment, do you think it makes sense to define a staging plan because I think it may take more time conclude the final solution and implementation?

- Stage 0: The runtime-related APIs to be implemented remain device-specific. Moreover, these APIs should meet actual requirements, rather than merely aiming for comprehensive coverage.
- Stage 1: Align runtime related python APIs with other devices

Are you okay with it? @albanD

File: torch/xpu/__init__.py
Comment (albanD) [inline torch/xpu/__init__.py:438]: nit: explicit imports are better for typing and readability. Would be nice to fix that going forward and remove the noqa

**Review by albanD — APPROVED** [review]
Feel free to cleanup the import in a follow up PR.

Sounds ok as a temporary solution (didn't carefully review the logic). I do expect that if the goal is to support HF and other third party, the unified API is even more urgent as otherwise it's going to be a never ending battle to update every single one of these libraries.

**Comment by EikanWang** [comment]
> I do expect that if the goal is to support HF and other third party, the unified API is even more urgent as otherwise it's going to be a never ending battle to update every single one of these libraries.

Thanks, @albanD . And I could not agree more.

File: torch/xpu/__init__.py
Comment (guangyey) [reply torch/xpu/__init__.py]: Thanks very much. I removed `noqa` in this PR.

**Comment by guangyey** [comment]
> Feel free to cleanup the import in a follow up PR.
> 
> Sounds ok as a temporary solution (didn't carefully review the logic). I do expect that if the goal is to support HF and other third party, the unified API is even more urgent as otherwise it's going to be a never ending battle to update every single one of these libraries.

Yes, I agree. I will prioritize and focus on these unified API tasks.

**Pull request closed without merging**
